lib/Sema/TypeCheckPattern.cpp

```diff
@@ -775,10 +775,31 @@ static bool validateParameterType(ParamDecl *decl, DeclContext *DC,
   return hadError;
 }
 
+/// Request nominal layout for any types that could be sources of typemetadata
+/// or conformances.
+void TypeChecker::requestRequiredNominalTypeLayoutForParameters(
+    ParameterList *PL) {
+  for (auto param : *PL) {
+    auto type = param->hasType() ? param->getType()->getCanonicalType()
+                                 : param->getTypeLoc().getType();
+    if (!type)
+      continue;
+    if (auto *generic = dyn_cast<BoundGenericType>(type.getPointer())) {
+      // Generic types are sources for typemetadata and conformances. If a
+      // parameter is of dependent type then the body of a function with said
+      // parameter could potentially require the generic type's layout to
+      // recover them.
+      if (auto *nominalDecl = dyn_cast<NominalTypeDecl>(generic->getDecl())) {
+        requestNominalLayout(nominalDecl);
+      }
+    }
+  }
+}
+
 /// Type check a parameter list.
-bool TypeChecker::typeCheckParameterList(ParameterList *PL, DeclContext *DC,
-                                         TypeResolutionOptions options,
-                                         GenericTypeResolver &resolver) {
+bool TypeChecker::typeCheckParameterList(
+    ParameterList *PL, DeclContext *DC, TypeResolutionOptions options,
+    GenericTypeResolver &resolver, bool bodyCouldRequireTypeOrConformance) {
```

Comment:
`typeCheckParameterList()` can get called for functions that are in non-primary source files (but which could have bodies), so this will require nominal layouts that aren't strictly needed. I recommend that the call to `typeCheckAbstractFunctionBody` go into `TypeChecker::typeCheckAbstractFunctionBody`, which is only called when we're type-checking the body itself. That more accurately represents when we'll need the layouts.